Clarify resolving implicit connections (3.1.1)

handrews opened the pull request:
Fixes:
* #2520 
* #3591 
* #3776 (sort-of, see the last "Reviewer Question" below)
* and various questions no one thought to ask in issues 🙃 

This clarifies how to handle resolving implicit (non-URI-based) connections in multi-document OpenAPI Descriptions.  While the behavior is implementation-defined overall, two approaches are RECOMMENDED for tool developers and OAD authors: one that is best suited for sharing components among OADs, and one that is compatible with pre-3.1 parsing approaches and is convenient for breaking a monolithic document up purely for author convenience.

Note that the term "complete OpenAPI document" has been defined in another change (PR #3820) pending approval on the 3.0.4 branch.

### Reviewer Questions

_This will be taken out of draft when I am confident that we have the right approach.  Also, this will get backported to 3.0.4, but without the full-document part; the backport will be easy once we've agreed on the 3.1 approach._

* For the resolve-within-complete documents approach, I decided that it was most consistent to treat all `operationId`s from all Path Item Objects in any of the documents to be available; we could instead still require that only those reachable via the entry document's `paths` or `webhooks` are considered (in any event, this PR recommends using `operationRef` instead as it has more predictable behavior)
* For the one-OpenAPI-Object-only approach, I went with Path Item Objects reachable through `paths` or `webhooks`, but we could make whole-document as well since all documents are fully parsed
* For the `allOf` usage of the Discriminator Object in both approachs, this makes any Schema Object able to participate in such an arrangement since they are all detectable by the full-document parsing requirement, just like `$id`, `$anchor`, etc. are detectable; it seems worth calling this out as it might be surprising
* The "for forward compatibility with 3.2" bit for the Security Requirement Object is my attempt at a Python-esque "from future import..." for a simple enhancement (see PR #3821) that is easily implemented to make shared component documents much more consistent in 3.1.  ***I am happy to split this into its own PR if it is at all controversial.*** Or even just take it out.  But it felt like such a vexing restriction, and it's easy to address without requiring full 3.2 support (whatever that ends up being).  The reason I added the opt-in part is that someone might rely on getting an error for something that is not a valid component name (and also you have to opt into Python's "from future" imports as well)

ralfhandl commented on versions/3.1.1.md:
```diff
@@ -182,12 +182,48 @@ It is the responsibility of an embedding format to define how to parse embedded
 
 If the same JSON/YAML object is parsed multiple times and the respective contexts require it to be parsed as _different_ Object types, the resulting behavior is _implementation defined_, and MAY be treated as an error if detected.  An example would be referencing an empty Schema Object under `#/components/schemas` where a Path Item Object is expected, as an empty object is valid for both types.   ...
 
+#### <a name="resolvingIndirectLinkages"></a>Resolving Indirect Connections
```
Why is the anchor not the camelCasedVersion of the section title?

handrews replied:
Because I changed the title a bunch.  Good catch!

ralfhandl commented on versions/3.1.1.md:
```diff
@@ -182,12 +182,48 @@ It is the responsibility of an embedding format to define how to parse embedded
+This requires ensuring that all referenced documents have been parsed prior to determining an `operationId` to be unresolvable.
+
+**Note:** Currently, Security Requirement Objects do not support a URI-based mechanism for connecting to Security Scheme Objects, which means that in this approach, they can only connect to Security Schemes within the same document.
+This limitation is expected to be addressed in the next minor version of this specification.
```
```suggestion
This limitation is expected to be addressed in a future version of this specification.
```
Don't over-promise.

handrews replied:
The PR for this is up for 3.2, so if we can agree on it, then "next minor" is correct and sets a better expectation.

handrews replied:
On second thought, it's better to be more vague as we can always add specifics later if we agree to the 3.2 change.  But I don't want us to have to think about 3.2 before doing these patch releases, so I'll commit this suggestion.

ralfhandl commented on versions/3.1.1.md:
```diff
@@ -182,12 +182,48 @@ It is the responsibility of an embedding format to define how to parse embedded
+The approach in this section is RECOMMENDED for OADs that are composed entirely of complete OpenAPI documents, optionally including standalone JSON Schema documents as long as any Discriminator Objects within them only use explicit `mapping` URIs for all possible values.
+This OAD design is RECOMMENDED when sharing components among multiple OADs.
+
+In this approach, each document is self-contained and resolves components to its own Components Object.
```
So I can't reference component schemas from another OAD object, I have to extract them into separate JSON Schema documents?

For whom would that be a simplification?

handrews replied:
No that's not what this says at all.  There is nothing here about forbidding Schema Objects in OpenAPI documents.  What is causing you to arrive at this interpretation?

ralfhandl reviewed (changes requested):
Need more explanation

handrews commented on versions/3.1.1.md:
```diff
@@ -182,12 +182,48 @@ It is the responsibility of an embedding format to define how to parse embedded
+The approach in this section is RECOMMENDED for OADs where the entry OpenAPI document is the only document containing an OpenAPI Object.
+This approach is most suitable for splitting an otherwise self-contained single-document OAD into multiple documents for author convenience.
+
+In this approach, all component names are resolved through the entry document's Components Object, and all `operationId`s are resolved through Path Item Objects reachable through the `paths` or `webhooks` fields.
```
There is nothing here about not being able to directly reference schemas.  This is about how component names (not URIs) are resolved.  Do you understand the distinction?  If not, let's figure out better wording.

handrews commented:
@ralfhandl I've updated the text in several places (it's a separate commit so you can look at the individual diff) - please let me know if that makes things more clear.  I appreciate your attention to detail here, because if the text confuses anyone on the TSC it will definitely confuse other readers who have much less context!

ralfhandl reviewed (dismissed):
Good clarification!

handrews commented:
I've converted this back to a draft because I realized I forgot tags, added them in the most recent commit, and then rethought the whole thing to have a single approach rather than two different ways to organize things.  Here is the alternative PR:

* #3856

ralfhandl commented:
I prefer the alternative proposal in #3856.

miqui commented:
@handrews , @ralfhandl  The language in #3856 is less confusing AFAIK (As a spec author (and trying to take a pragmatic approach)).

handrews commented:
Closing as we accepted #3856 instead 🎉
